File: 39. 组合总和/Solution.py
class Solution:
    # 剪枝：如果下一层的sum（就是本层的 sum + candidates[i]）已经大于target，就可以结束本轮for循环的遍历。
    def traversal(self, candidates, target, path, ans, Sum, start):
        if Sum == target:
            ans.append(path[:])
        if Sum > target:
            return
        for i in range(start, len(candidates)):
            if Sum + candidates[i] > target: # 剪枝
                break
            Sum += candidates[i]
            path.append(candidates[i])
            self.traversal(candidates, target, path, ans, Sum, i) # 不用i+1了，表示可以重复读取当前的数
            Sum -= candidates[i]
            path.pop()

    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        ans = []
        path = []
        candidates.sort()
        self.traversal(candidates, target, path, ans, 0, 0)
        return ans

File: 39. 组合总和/Solution.cpp
#include<bits/stdc++.h>

using namespace std;

class Solution {
private:
    vector<vector<int>> result;
    vector<int> path;
    void backtracking(vector<int>& candidates, int target, int sum, int startIndex) {
        if (sum == target) {
            result.push_back(path);
            return;
        }

        // 如果 sum + candidates[i] > target 就终止遍历
        for (int i = startIndex; i < candidates.size() && sum + candidates[i] <= target; i++) {
            sum += candidates[i];
            path.push_back(candidates[i]);
            backtracking(candidates, target, sum, i);
            sum -= candidates[i];
            path.pop_back();

        }
    }
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        result.clear();
        path.clear();
        sort(candidates.begin(), candidates.end()); // 需要排序
        backtracking(candidates, target, 0, 0);
        return result;
    }
};
